console/v1/types_console_sample.go

```diff
@@ -0,0 +1,166 @@
+	// +optional
+	Badge string `json:"badge"`
+
+	// tags like "Java", "Quarkus", etc.
+	// Tags are optional and can be used by the user to find samples in the samples catalog.
+	// They will be also displayed on the samples details page.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Tags []string `json:"tags"`
+
+	// The actual sample source, like an external git repository or container image.
+	// +kubebuilder:validation:Required
+	// +required
+	Source ConsoleSampleSource `json:"source"`
+}
+
+// ConsoleSampleSourceType is an enumeration of the supported sample types.
+// Unsupported samples types will be ignored in the web console.
+// +kubebuilder:validation:Enum:=GitImport;ContainerImport
+type ConsoleSampleSourceType string
+
+const (
+	// A sample that let the user import code from a git repository.
+	GitImport ConsoleSampleSourceType = "GitImport"
+	// A sample that let the user import a container image.
+	ContainerImport ConsoleSampleSourceType = "ContainerImport"
+)
+
+// ConsoleSampleSource is the actual sample definition and can hold different sample types.
+// Unsupported sample types will be ignored in the web console.
+// +union
+// +kubebuilder:validation:XValidation:rule="self.type == 'GitImport' ? has(self.gitImport) : !has(self.gitImport)",message="source.gitImport is required when source.type is GitImport, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="self.type == 'ContainerImport' ? has(self.containerImport) : !has(self.containerImport)",message="source.containerImport is required when source.type is ContainerImport, and forbidden otherwise"
+type ConsoleSampleSource struct {
+	// type of the sample, currently supported: "GitImport";"ContainerImport"
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:="GitImport";"ContainerImport"
+	// +kubebuilder:validation:Required
+	Type ConsoleSampleSourceType `json:"type"`
+
+	// gitimport let the user import code from a git repository.
+	// +unionMember,optional
+	GitImport *ConsoleGitImportSample `json:"gitImport,omitempty"`
+
+	// containerimport let the user import a container image.
+	// +unionMember,optional
+	ContainerImport *ConsoleContainerImportSample `json:"containerImport,omitempty"`
+}
+
+// ConsoleGitImportSample let the user import code from a public git repository.
+type ConsoleGitImportSample struct {
+	// Git repository URL.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	URL string `json:"url"`
+	// Optional branch, tag, or commit.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Ref string `json:"ref"`
+}
+
+// ConsoleContainerImportSample let the user import a container image.
+type ConsoleContainerImportSample struct {
+	// Reference to a container image like openshift/hello-openshift,
+	// docker.io/openshift/hello-openshift or quay.io/<username>/<image-name>
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
```

Comment:
Can we put in some arbitrarily high maximum just in case, eg 2048?

Reply:
Added a MaxLength for the `source.git.url` and `source.image`.